File: src/main/scala/de/awagen/kolibri/datatypes/collections/generators/IndexedGenerator.scala
/**
  * Copyright 2021 Andreas Wagenmann
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

package de.awagen.kolibri.datatypes.collections.generators

import de.awagen.kolibri.datatypes.io.KolibriSerializable
import de.awagen.kolibri.datatypes.types.SerializableCallable.SerializableFunction1

/**
  * Generator providing generation logic elements of given type. Also allows splitting into smaller part given
  * start and end index. Provides method to get nr of elements, elements at position i and finally provides
  * an iterator over the elements of type T. Note that the idea here is that the objects generated by the generator
  * dont need to be in memory, but can be generated by the logic described within the iterator. Also, determining size
  * here should not require all elements to be loaded.
  *
  * Example use case: given Seq of Seq of different parameter values, find the i-th permutation of the parameter
  * set. If we have many parameter combinations, it would be memory-inefficient storing them all instead of providing logic
  * to generate on demand
  *
  * NOTE: this class could extend Iterable[T], yet this seems to cause
  * some problems with Kryo Serialization when assigning KolibriSerializable
  * to kryo serializer (e.g correct type of generator is not preserved);
  * should be solvable by proper config. Until then dont extend Iterable
  * (otherwise should see serialization log error that desired interface is not implemented but element being of
  * type scala.collection.immutable.$colon$colon)
  *
  * @tparam T
  */
trait IndexedGenerator[+T] extends KolibriSerializable {

  val nrOfElements: Int

  /**
    * Iterator over contained elements
    *
    * @return
    */
  def iterator: Iterator[T] = {
    new Iterator[T] {
      var currentPosition = 0

      override def hasNext: Boolean = currentPosition < nrOfElements

      override def next(): T = {
        val element = get(currentPosition)
        currentPosition += 1
        element.get
      }
    }
  }

  def size: Int = nrOfElements

  /**
    * create generator that only generates a part of the original generator.
    *
    * @param startIndex : startIndex (inclusive)
    * @param endIndex   : endIndex (exclusive)
    * @return generator generating the subpart of the generator as given by startIndex and endIndex
    */
  def getPart(startIndex: Int, endIndex: Int): IndexedGenerator[T]

  /**
    * Get the index-th element
    *
    * @param index
    * @return
    */
  def get(index: Int): Option[T]

  /**
    * Provided a mapping function, create generator of new type where elements are created by current generator
    * and then mapped by the provided function
    *
    * @param f : mapping function
    * @tparam B : the type the original element type is mapped to
    * @return : new generator providing the new type
    */
  def mapGen[B](f: SerializableFunction1[T, B]): IndexedGenerator[B]

}

File: kolibri-fleet/src/lib/values/generators/indexed_generator.py
# Copyright 2023 Andreas Wagenmann
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from abc import ABC, abstractmethod
from typing import TypeVar, Iterator, Optional, Callable, List, Generic

T = TypeVar('T')
U = TypeVar('U')


class IndexedGenerator(ABC, Generic[T]):

    class ElementIterator(Iterator):

        def __init__(self, generator: 'IndexedGenerator[T]'):
            self.generator = generator
            self.current_position = 0

        def has_next(self):
            return self.current_position < self.generator.size

        def __next__(self):
            if not self.has_next():
                raise StopIteration
            element = self.generator.get(self.current_position)
            self.current_position += 1
            return element

    def iterator(self) -> Iterator[T]:
        """
        Iterator over contained elements
        :return:
        """
        return IndexedGenerator.ElementIterator(self)

    @property
    @abstractmethod
    def size(self) -> int:
        pass

    @abstractmethod
    def get(self, index: int) -> T | None:
        """
        Get the index-th element
        :param index:
        :return:
        """
        pass

    @abstractmethod
    def get_part(self, start_index: int, end_index: int) -> 'IndexedGenerator[T]':
        """
        create generator that only generates a part of the original generator.
        :param start_index: startIndex (inclusive)
        :param end_index: endIndex (exclusive)
        :return: generator generating the subpart of the generator as given by startIndex and endIndex
        """
        pass

    @abstractmethod
    def map_gen(self, func: Callable[[T], U]) -> 'IndexedGenerator[U]':
        """
        Provided a mapping function, create generator of new type where elements are created by current generator
        and then mapped by the provided function
        :param func: mapping function
        :return: new generator providing the new type
        """
        pass


class ByFunctionIndexedGenerator(IndexedGenerator[T]):

    def __init__(self, nr_of_elements: int, generator_function: Callable[[int], T]):
        self.nr_of_elements = nr_of_elements
        self.generator_function = generator_function

    @staticmethod
    def create_from_list(data: List[T]):
        return ByFunctionIndexedGenerator(len(data), lambda x: data[x])

    @property
    def size(self) -> int:
        return self.nr_of_elements

    def get_part(self, start_index: int, end_index: int) -> 'IndexedGenerator[T]':
        assert(start_index > 0)
        end = min(self.size, end_index)
        new_size = end - start_index
        return ByFunctionIndexedGenerator(new_size, lambda x: self.get(x + start_index))

    def get(self, index: int) -> Optional[T]:
        if index < self.size:
            return self.generator_function(index)
        return None

    def map_gen(self, func: Callable[[T], U]) -> 'IndexedGenerator[U]':
        def combined_callable() -> Callable[[int], U]:
            return lambda x: func(self.generator_function(x))
        return ByFunctionIndexedGenerator(self.size, combined_callable())
